### PR TITLE
Rework `Var` to leverage `Initializer()`

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:

**NOTE:** this PR is an extension of PR #2172 

This PR moves `Var` initialization to leverage the `Initializer` mechanism for processing construction arguments (thereby centralizing / standardizing how we process constants / rules / arguments).

It also updates / simplifies the logic for casting NumericValues. (i.e., through `__bool__`, `__float__`, and `__int__`).  Previous PRs updated relational expressions to generate an exception when a non-constant expression was cast to `bool` (i.e., in an `if`).  This PR promulgates that into the rest of `NumericValue`.  This fixes an issue where 
```python
m.x = Var()
if m.x != 0:
    pass
```
would generate an exception, but
```python
m.x = Var()
if m.x:
    pass
```
would not.  In addition, this PR removes (unnecessary) re-implementations of these methods throughout `pyomo.core.base`.

Finally, it centralizes processing of bounds and variable value validation.  This adds checking for the variable bounds, but does not emit a warning (pending the final decision on #2180).  Despite the additional overhead of checking bounds, overall performance is still improves (about 5% reduction in model construction time, 1% overall across the entire performance test suite).  

Comparing main(79fb168) against this PR (e30260f):

image

Note that the slight performance degradation in the GAMS writer will be resolved in a subsequent PR.

## Changes proposed in this PR:
- Overhaul `Var`
  - Leverage `Initializer()` in component construction
  - Leverage `@disable_methods()` (`AbstractScalarVar`) to perform error checking on unconstructed `ScalarVar` objects
  - Centralize incoming value validation within `set_value()`
  - Add bounds checking in `set_value()`
- `NumericValue.__bool__`, `__float__`, and `__int__` now always raise an exception for non-constant `NumericValue` instances
- remove reimplementations of `__float__` and `__int__`
- minor efficiency improvement to `Component.__setstate__`
- minor improvements to Set efficiency and error messages
- remove `__nonzero__` (a Python 2-ism)

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/base/var.py [hunk 1]
```diff
@@ -371,70 +354,169 @@ def __getstate__(self):
...
+                _src_units = units.get_units(val)
+                val = units.convert_value(
+                    num_value=_src_magnitude, from_units=_src_units,
+                    to_units=self.parent_component()._units)
+            else:
+                val = value(val)
+
+        if not valid:
+            if val not in self.domain:
+                # logger.warning(
```
Comment (mrmundt) [inline pyomo/core/base/var.py:391]: Do we need this commented out section?

File: pyomo/core/base/var.py [hunk 2]
```diff
@@ -371,70 +354,169 @@ def __getstate__(self):
...
+                raise ValueError("Numeric value `%s` (%s) is not in "
+                                 "domain %s for Var %s" %
+                                 (val, type(val).__name__,
+                                  self.domain, self.name))
+            elif (self._lb is not None and val < value(self._lb)) or (
+                    self._ub is not None and val > value(self._ub)):
+                pass
+                # logger.warning(
```
Comment (mrmundt) [inline pyomo/core/base/var.py:399]: Was the intent to pass or log a warning?

File: pyomo/core/base/var.py [hunk 3]
```diff
@@ -612,47 +671,124 @@ def get_units(self):
...
+                # initializers that are constant, we can avoid
+                # re-calling (and re-validating) the inputs in certain
+                # cases.  To support this, we will create the first
+                # _VarData and then use it as a template to initialize
+                # (constant portions of) every VarData so as to not
+                # repeat all the domain/bounds validation.
+                try:
+                    ref = self._getitem_when_not_present(
+                        next(iter(self.index_set())))
+                    # state = ref.__getstate__()
```
Comment (mrmundt) [inline pyomo/core/base/var.py:743]: Do these lines still need to be here?

File: pyomo/core/tests/unit/test_numeric_expr.py [hunk 4]
```diff
@@ -2691,12 +2699,13 @@ def checkCondition(self, expr, expectedValue, use_value=False):
     def test_immutable_paramConditional(self):
         model = AbstractModel()
         model.p = Param(initialize=1.0, mutable=False)
-        #
-        try:
+        # Immutable Params appear mutable (nonconstant) before they are
```
Comment (mrmundt) [inline pyomo/core/tests/unit/test_numeric_expr.py:2702]: For consistency: `non-constant`

File: pyomo/core/tests/unit/test_numeric_expr.py [hunk 5]
```diff
@@ -2731,18 +2740,32 @@ def test_immutable_paramConditional(self):
     def test_immutable_paramConditional_reversed(self):
         model = AbstractModel()
         model.p = Param(initialize=1.0, mutable=False)
-        #
-        # TODO: Inequalities evaluate True when the parameter is unconstructed?
-        #
-        with self.assertRaises(PyomoException):
+        # Immutable Params appear mutable (nonconstant) before they are
```
Comment (mrmundt) [inline pyomo/core/tests/unit/test_numeric_expr.py:2743]: For consistency: `non-constant`

File: pyomo/core/tests/unit/test_numeric_expr.py [hunk 6]
```diff
@@ -2771,12 +2794,13 @@ def test_immutable_paramConditional_reversed(self):
     def test_immutable_paramConditional_reversed(self):
         model = AbstractModel()
         model.p = Param(initialize=1.0, mutable=False)
-        #
-        try:
+        # Immutable Params appear mutable (nonconstant) before they are
```
Comment (mrmundt) [inline pyomo/core/tests/unit/test_numeric_expr.py:2797]: For consistency: `non-constant`

**Review by mrmundt — CHANGES_REQUESTED** [review]

File: pyomo/core/base/var.py (hunk 1 above)
Comment (jsiirola) [reply pyomo/core/base/var.py:387]: For now, yes: this is holding the (hopeful) implementation of #2180.

File: pyomo/core/base/var.py (hunk 2 above)
Comment (jsiirola) [reply pyomo/core/base/var.py:395]: For now, pass.  Pending #2180, a warning.

**Review by mrmundt — APPROVED** [review]

**Pull request merged**